### PR TITLE
feat: add gzip support for envd file upload and download

### PR DESCRIPTION
**Pull request by mishushakov**
## Summary

Add gzip compression support for envd file transfers:
- Clients can send gzip-compressed requests via `Content-Encoding: gzip` header
- Server responds with gzip compression when client sends `Accept-Encoding: gzip`
- Unsupported encodings are rejected with 400 Bad Request
- Extensible design to support additional encodings in the future

## Implementation

- `encoding.go`: Shared validation and compression functions with `SupportedEncodings` list
- `encoding_test.go`: Comprehensive test coverage for encoding validation
- Updated `upload.go` and `download.go` to use the encoding functions

🤖 Generated with Claude Code

---

> [!NOTE]
> **Medium Risk**
> Touches HTTP request/response streaming for file upload/download; incorrect header negotiation or gzip handling could break range requests, caching, or cause corrupted transfers under certain clients.
> 
> **Overview**
> Adds gzip compression support for envd file transfers: uploads can now accept `Content-Encoding: gzip` bodies (decompressed before multipart parsing), and downloads negotiate `Accept-Encoding` to optionally return `Content-Encoding: gzip` with `Vary: Accept-Encoding`, while rejecting unacceptable encodings with `406` and preserving `http.ServeContent` semantics for range/conditional requests by forcing identity.
> 
> Introduces shared encoding parsing/validation utilities and tests, expands API tests to cover gzip upload/download and the range/identity edge case, and bumps envd `Version` to `0.5.3`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fa7e379ba8108774a1b6d3b27044bc5b6f972be1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/envd/internal/api/upload.go [hunk 1]
```diff
@@ -193,6 +193,20 @@ func (a *API) PostFiles(w http.ResponseWriter, r *http.Request, params PostFiles
 		l.Msg("File write")
 	}()
 
+	// Handle gzip-encoded request body
+	body, err := getDecompressedBody(r)
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/envd/internal/api/upload.go:229]: I think here it is okay

**Comment by mishushakov** [comment]
@djeebus let me know if we want to move the encoding utils to shared

File: packages/envd/internal/api/download.go [hunk 2]
```diff
@@ -101,6 +103,29 @@ func (a *API) GetFiles(w http.ResponseWriter, r *http.Request, params GetFilesPa
...
+
+	// Disable gzip for Range or conditional requests to preserve http.ServeContent behavior
+	// (206 Partial Content, 304 Not Modified)
+	hasRangeOrConditional := r.Header.Get("Range") != "" ||
+		r.Header.Get("If-Modified-Since") != "" ||
+		r.Header.Get("If-None-Match") != "" ||
+		r.Header.Get("If-Range") != ""
+	if hasRangeOrConditional {
+		encoding = ""
+	}
```
Comment (mishushakov) [inline packages/envd/internal/api/download.go]: not 100% sure about this, I just didn't want to re-implement the current behavior with gzip

File: packages/envd/internal/api/encoding.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		if eq.encoding == "identity" || eq.encoding == "*" {
+			return "", nil
+		}
+
+		if isSupportedEncoding(eq.encoding) {
+			return eq.encoding, nil
+		}
+	}
+
+	return "", fmt.Errorf("unsupported Accept-Encoding: %s, supported: %v", header, SupportedEncodings)
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/envd/internal/api/encoding.go]: Imo I think we should keep error if you're using unsupported encoding

File: packages/envd/internal/api/download.go (hunk 2 above)
Comment (mishushakov) [reply packages/envd/internal/api/download.go]: I can throw 400 if we do gzip and any of the headers?

File: packages/envd/internal/api/download.go [hunk 4]
```diff
@@ -101,6 +104,29 @@ func (a *API) GetFiles(w http.ResponseWriter, r *http.Request, params GetFilesPa
 		return
 	}
 
+	// Validate Accept-Encoding header
+	encoding, err := parseAcceptEncoding(r)
+	if err != nil {
+		errMsg = err
```
Comment (jakubno) [inline packages/envd/internal/api/download.go]:
```suggestion
		errMsg = fmt.Errorf("error while parsing encoding: %w", err)
```

File: packages/envd/internal/api/download.go (hunk 2 above)
Comment (jakubno) [reply packages/envd/internal/api/download.go]: Sounds cleaner to me

File: packages/envd/internal/api/download.go [hunk 5]
```diff
@@ -111,5 +137,31 @@ func (a *API) GetFiles(w http.ResponseWriter, r *http.Request, params GetFilesPa
 	}
 	defer file.Close()
 
-	http.ServeContent(w, r, path, time.Now(), file)
+	// Serve with gzip encoding if requested and no Range/conditional headers.
+	// Note: If io.Copy fails after headers are sent, the client receives a truncated
+	// gzip stream with HTTP 200. Buffering the entire response would fix this but
+	// has memory implications for large files. Clients should validate gzip integrity.
+	if encoding == "gzip" {
```
Comment (jakubno) [inline packages/envd/internal/api/download.go]: can we make `gzip` a constant?

File: packages/envd/internal/api/download.go [hunk 6]
```diff
@@ -101,6 +104,29 @@ func (a *API) GetFiles(w http.ResponseWriter, r *http.Request, params GetFilesPa
...
+
+		return
+	}
+
+	// Tell caches to store separate variants for different Accept-Encoding values
+	w.Header().Set("Vary", "Accept-Encoding")
+
+	// Disable gzip for Range or conditional requests to preserve http.ServeContent behavior
+	// (206 Partial Content, 304 Not Modified)
+	hasRangeOrConditional := r.Header.Get("Range") != "" ||
```
Comment (jakubno) [inline packages/envd/internal/api/download.go:123]: we need to check this only for gzip, right?

File: packages/envd/internal/api/encoding.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakubno) [reply packages/envd/internal/api/encoding.go]: I think you should comply with RFC, you should default to identity, if not explicitly forbidden

File: packages/envd/internal/api/upload.go [hunk 7]
```diff
@@ -149,8 +149,6 @@ func resolvePath(part *multipart.Part, paths *UploadSuccess, u *user.User, defau
 }
 
 func (a *API) PostFiles(w http.ResponseWriter, r *http.Request, params PostFilesParams) {
-	defer r.Body.Close()
```
Comment (jakubno) [inline packages/envd/internal/api/upload.go:182]: what if e.g. validateSigning returns an error, then you'll be leaking

File: packages/envd/internal/api/download.go (hunk 6 above)
Comment (mishushakov) [reply packages/envd/internal/api/download.go:123]: any encoding

File: packages/envd/internal/api/download.go [hunk 8]
```diff
@@ -117,8 +117,9 @@ func (a *API) GetFiles(w http.ResponseWriter, r *http.Request, params GetFilesPa
 	// Tell caches to store separate variants for different Accept-Encoding values
 	w.Header().Set("Vary", "Accept-Encoding")
 
-	// Disable gzip for Range or conditional requests to preserve http.ServeContent behavior
-	// (206 Partial Content, 304 Not Modified)
+	// Fall back to identity for Range or conditional requests to preserve http.ServeContent
+	// behavior (206 Partial Content, 304 Not Modified). Accept-Encoding is advisory per HTTP
+	// spec, so the server can always choose to ignore it.
```
Comment (jakubno) [inline packages/envd/internal/api/download.go]: we could pass `hasRangeOrConditional` to `parseAcceptEncoding` so we can handle the case with forbidden identity?

File: packages/envd/internal/api/encoding.go [hunk 9]
```diff
@@ -106,15 +108,34 @@ func parseAcceptEncoding(r *http.Request) (string, error) {
...
 	for _, eq := range encodings {
 		// Skip encodings with q=0 (explicitly rejected)
 		if eq.quality == 0 {
 			continue
 		}
 
-		// "identity" means no encoding, "*" means any encoding is acceptable
-		if eq.encoding == "identity" || eq.encoding == "*" {
+		// "identity" means no encoding
+		if eq.encoding == "identity" {
```
Comment (jakubno) [inline packages/envd/internal/api/encoding.go]: can you also make `identity` a const 🙏🏻

File: packages/envd/internal/api/download.go [hunk 10]
```diff
@@ -101,6 +104,30 @@ func (a *API) GetFiles(w http.ResponseWriter, r *http.Request, params GetFilesPa
...
+
+	// Fall back to identity for Range or conditional requests to preserve http.ServeContent
+	// behavior (206 Partial Content, 304 Not Modified). Accept-Encoding is advisory per HTTP
+	// spec, so the server can always choose to ignore it.
+	hasRangeOrConditional := r.Header.Get("Range") != "" ||
+		r.Header.Get("If-Modified-Since") != "" ||
+		r.Header.Get("If-None-Match") != "" ||
+		r.Header.Get("If-Range") != ""
+	if hasRangeOrConditional {
+		encoding = ""
```
Comment (jakubno) [inline packages/envd/internal/api/download.go]: I would prefer to return Identity instead of `""`, it will make it more readable

File: packages/envd/internal/api/download.go (hunk 8 above)
Comment (mishushakov) [reply packages/envd/internal/api/download.go]: i think it's fine to leave as is, more explicit what is happening

File: packages/envd/internal/api/encoding.go [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to create gzip reader: %w", err)
+		}
+
+		return gzReader, nil
+	default:
+		// This shouldn't happen if isSupportedEncoding is correct
+		return nil, fmt.Errorf("encoding %s is supported but not implemented", encoding)
+	}
+}
```
[the remark this answers is not in the document]
Comment (mishushakov) [reply packages/envd/internal/api/encoding.go:227]: it's inside their own sandboxes

**Review by jakubno — APPROVED** [review]

**Review by djeebus — APPROVED** [review]

File: packages/envd/internal/api/download_test.go [hunk 12]
```diff
@@ -163,3 +163,44 @@ func TestGetFilesContentDispositionWithNestedPath(t *testing.T) {
 	contentDisposition := resp.Header.Get("Content-Disposition")
 	assert.Equal(t, `inline; filename=document.pdf`, contentDisposition, "Content-Disposition should contain only the filename, not the path")
 }
+
+func TestGetFiles_GzipEncoding_ExplicitIdentityOffWithRange(t *testing.T) {
```
Comment (djeebus) [inline packages/envd/internal/api/download_test.go:171]: It would be good to write a test that actually downloads a compressed file.
